File: src/UniCaffeWeb/classes/dao/DAO.php
<?php
/**
 * Faz conexão com banco de dados e gerencia persistências.
 * @author Jefferson Uchôa Ponte
 *
 */
class DAO {
    /**
     * @var string
     * Path do arquivo de configuração do banco de dados. 
     */
	const ARQUIVO_CONFIGURACAO = "/dados/unicaffe/unicaffe_bd.ini";
	/**
	 * @var integer
	 * Indica conexão com o UniCaffé Servidor. 
	 */
	const TIPO_UNICAFFE = 0;
	/**
	 * @var integer
	 * Indica conexão com base de dados padrão. 
	 */
	const TIPO_DEFAULT = 1;
	/**
	 * @var integer
	 * Indica conexão com base de dados de usuários. 
	 */
	const TIPO_USUARIOS = 2;
	/**
	 * @var integer
	 * Indica conexão de base de dados para autenticação. 
	 */
	const TIPO_AUTENTICACAO = 3;
	/**
	 * @var mixed 
	 */
	protected $conexao;
	/**
	 * @var int
	 */
	private $tipoDeConexao;
	/**
	 * @var string sistema gerenciador de banco de dados.
	 */
	private $sgbb;
	/**
	 * @var string entidade para autenticação ou busca de usuários. 
	 */
	private $entidade;
	
	/**
	 * @return string
	 */
	public function getEntidade(){
		return $this->entidade;
	}
	/**
	 * Obtém objeto DAO. Atribua uma conexão no primeiro parâmetro caso queira reaproveitá-la.
	 * Caso queira abrir uma nova conexão diferente da conexão do banco de dados padrão atribua null
	 * no primeiro parâmetro e no segundo atribua uma constante da classe DAO para definir o tipo
	 * de conexão.
	 *
	 *
	 * @param PDO|UniCaffe $conexao
	 * @param integer $tipo
	 */
	public function __construct($conexao = null, $tipo = self::TIPO_DEFAULT) {
		$this->tipoDeConexao = $tipo;
		if ($conexao != null) {
			$this->conexao = $conexao;
		} else {
			$this->fazerConexao ();
		}
	}
	/**
	 * Faz uma conexão com banco de dados ou com o UniCaffé.
	 *  A biblioteca UniCaffe.php trata conexão com banco de dados com o servidor do UniCaffé
	 *  de forma muito semelhante à forma usada pelo PDO. Por esse motivo foi adicionado aqui como 
	 *  um tipo de conexão do DAO. 
	 */
	public function fazerConexao() {
		$config = parse_ini_file ( self::ARQUIVO_CONFIGURACAO );
		
		switch ($this->tipoDeConexao) {
			case self::TIPO_UNICAFFE:
				$bd ['sgdb'] = "unicaffe";
				break;
			case self::TIPO_USUARIOS :
				$bd ['sgdb'] = $config ['usuarios_sgdb'];
				$bd ['nome'] = $config ['usuarios_bd_nome'];
				$bd ['host'] = $config ['usuarios_host'];
				$bd ['porta'] = $config ['usuarios_porta'];
				$bd ['usuario'] = $config ['usuarios_usuario'];
				$bd ['senha'] = $config ['usuarios_senha'];
				$this->entidade = $config['usuarios_entidade_nome'];
				break;
			case self::TIPO_AUTENTICACAO:
				$bd ['sgdb'] = $config ['autenticacao_sgdb'];
				$bd ['nome'] = $config ['autenticacao_bd_nome'];
				$bd ['host'] = $config ['autenticacao_host'];
				$bd ['porta'] = $config ['autenticacao_porta'];
				$bd ['usuario'] = $config ['autenticacao_usuario'];
				$bd ['senha'] = $config ['autenticacao_senha'];
				$this->entidade = $config['autenticacao_entidade_nome'];
				break;
			default :
				$bd ['sgdb'] = $config ['default_sgdb'];
				$bd ['nome'] = $config ['default_bd_nome'];
				$bd ['host'] = $config ['default_host'];
				$bd ['porta'] = $config ['default_porta'];
				$bd ['usuario'] = $config ['default_usuario'];
				$bd ['senha'] = $config ['default_senha'];
		}
	

		if ($bd ['sgdb'] == "postgres") {
			
			$this->conexao = new PDO ( 'pgsql:host=' . $bd ['host'] . ' dbname=' . $bd ['nome'] . ' user=' . $bd ['usuario'] . ' password=' . $bd ['senha'] );
		} else if ($bd ['sgdb'] == "mssql") {
			$this->conexao = new PDO ( 'dblib:host=' . $bd ['host'] . ';dbname=' . $bd ['nome'], $bd ['usuario'], $bd ['senha'] );
		} else if ($bd ['sgdb'] == "sqlite") {
			$this->conexao = new PDO ( 'sqlite:' . $bd ['nome'] );
			
		}else if($bd['sgdb'] == "unicaffe"){
			$this->conexao = new UniCaffe($config ['unicaffe_host'], $config ['unicaffe_porta']);
		}
		$this->sgdb = $bd ['sgdb'];
	}
	public function setConexao($conexao) {
		$this->conexao = $conexao;
	}
	public function getConexao() {
		return $this->conexao;
	}
	public function fechaConexao() {
		$this->conexao = null;
	}
	public function getTipoDeConexao() {
		return $this->tipoDeConexao;
	}
	public function setTipoDeConexao($tipo) {
		$this->tipoDeConexao = $tipo;
	}
	public static function criarBanco() {
		$dao = new DAO ();
		$sql[] = "CREATE TABLE usuarios_unicafe( id_usuario serial NOT NULL, nome character varying(200), email character varying(200), login character varying(200), senha character varying(200), CONSTRAINT usuarios_unicafe_pkey PRIMARY KEY (id_usuario))";
		$sql[] = "CREATE TABLE maquina( id_maquina serial NOT NULL, nome_pc character varying(100), mac character varying(200), CONSTRAINT maquina_pkey PRIMARY KEY (id_maquina));";
		$sql[] = "CREATE TABLE usuario( nome character varying(200), email character varying(200), login character varying(200), senha character varying(200), nivel_acesso integer, id_base_externa integer, id_usuario serial NOT NULL, CONSTRAINT usuario_pkey PRIMARY KEY (id_usuario));";
		$sql[] = "CREATE TABLE laboratorio( id_laboratorio serial NOT NULL, nome_laboratorio character varying(200), CONSTRAINT laboratorio_pkey PRIMARY KEY (id_laboratorio));";
		$sql[] = "CREATE TABLE laboratorio_maquina( id_laboratorio_maquina serial NOT NULL, id_maquina integer, id_laboratorio integer, CONSTRAINT laboratorio_maquina_pkey PRIMARY KEY (id_laboratorio_maquina), CONSTRAINT laboratorio_maquina_id_laboratorio_fkey FOREIGN KEY (id_laboratorio) REFERENCES laboratorio (id_laboratorio) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION, CONSTRAINT laboratorio_maquina_id_maquina_fkey FOREIGN KEY (id_maquina) REFERENCES maquina (id_maquina) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION);";
		$sql[] = "CREATE TABLE administrador( id_administrador serial NOT NULL, id_usuario integer, id_laboratorio integer, CONSTRAINT administrador_pkey PRIMARY KEY (id_administrador), CONSTRAINT administrador_id_laboratorio_fkey FOREIGN KEY (id_laboratorio) REFERENCES laboratorio (id_laboratorio) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION, CONSTRAINT administrador_id_usuario_fkey FOREIGN KEY (id_usuario) REFERENCES usuario (id_usuario) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION);";
		$sql[] = "CREATE TABLE acesso( id_acesso serial NOT NULL, hora_inicial timestamp without time zone, tempo_oferecido integer, tempo_usado integer, ip character varying(200), id_usuario integer, id_maquina integer, CONSTRAINT acesso_pkey PRIMARY KEY (id_acesso), CONSTRAINT acesso_id_maquina_fkey FOREIGN KEY (id_maquina) REFERENCES maquina (id_maquina) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION, CONSTRAINT acesso_id_usuario_fkey FOREIGN KEY (id_usuario) REFERENCES usuario (id_usuario) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION);";
		foreach ( $sql as $strSql ) {
			$dao->getConexao ()->query ( $strSql );
		}
	}
}

?>

File: src/UnicaffeServidor/src/main/java/br/edu/unilab/unicaffe/dao/DAO.java
package br.edu.unilab.unicaffe.dao;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

/**
 * Faz conexão com banco de dados e gerencia persistências. 
 * @author Jefferson Uchôa Ponte
 *
 */
public class DAO {

	/**
	 * Tipo de conexão. 
	 */
	private int tipoDeConexao;
	/**
	 * Sistema gerenciador de banco de dados. 
	 */
	private String sgdb;
	/**
	 * Tabela ou View de usuarios da base remota. 
	 */
	private String entidade;
	/**
	 * Conexão com banco. 
	 */
	private Connection conexao;

	/**
	 * Constroi objeto DAO com conexão com banco de dados. 
	 * @param tipoDeConexao
	 */
	public DAO(int tipoDeConexao) {
		this.tipoDeConexao = tipoDeConexao;
		fazerConexao();
	}

	/**
	 * Constroi objeto DAO com conexão com banco de dados.
	 */
	public DAO() {
		this.tipoDeConexao = TIPO_DEFAULT;
		fazerConexao();
	}

	/**
	 * Constroi objeto DAO com conexão com banco de dados.
	 */
	public DAO(Connection conexao) {
		this.conexao = conexao;
	}


	/**
	 * Faz uma conexão com banco de dados de acordo com as informações do arquivo de configuração. 
	 */
	public void fazerConexao() {
		this.conexao = null;
		try {
			Properties config = new Properties();
			FileInputStream file;
			file = new FileInputStream(ARQUIVO_CONFIGURACAO);
			config.load(file);
			String sgdb, host, porta, bdNome, usuario, senha;
			switch (tipoDeConexao) {
			case TIPO_USUARIOS:
				sgdb = config.getProperty("usuarios_sgdb");
				host = config.getProperty("usuarios_host");
				porta = config.getProperty("usuarios_porta");
				bdNome = config.getProperty("usuarios_bd_nome");
				usuario = config.getProperty("usuarios_usuario");
				senha = config.getProperty("usuarios_senha");
				this.entidade = config.getProperty("usuarios_entidade_nome");
				break;
			case TIPO_AUTENTICACAO:
				sgdb = config.getProperty("autenticacao_sgdb");
				host = config.getProperty("autenticacao_host");
				porta = config.getProperty("autenticacao_porta");
				bdNome = config.getProperty("autenticacao_bd_nome");
				usuario = config.getProperty("autenticacao_usuario");
				senha = config.getProperty("autenticacao_senha");
				senha = senha.replace("\"", "");
				this.entidade = config.getProperty("autenticacao_entidade_nome");
				break;
			default:
				sgdb = config.getProperty("default_sgdb");
				host = config.getProperty("default_host");
				porta = config.getProperty("default_porta");
				bdNome = config.getProperty("default_bd_nome");
				usuario = config.getProperty("default_usuario");
				senha = config.getProperty("default_senha");
				break;
			}
			file.close();
			if (sgdb.equals("postgres")) {
				Class.forName(DRIVER_POSTGRES);
				this.conexao = DriverManager.getConnection(JDBC_BANCO_POSTGRES+ "//" + host + "/" + bdNome, usuario, senha);
				
			} else if (sgdb.equals("sqlite")) {
				Class.forName(DRIVER_SQLITE);
				this.conexao = DriverManager.getConnection(JDBC_BANCO_SQLITE+bdNome);
			} else if (sgdb.equals("mysql")) {
				Class.forName(DRIVER_MYSQL);
				this.conexao = DriverManager.getConnection(JDBC_BANCO_MYSQL + "//" + host +":"+ porta + "/" + bdNome, usuario, senha);
			}

		} catch (ClassNotFoundException e1) {
			e1.printStackTrace();
		} catch (SQLException e) {
			e.printStackTrace();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Retorna a conexão com banco de dados. 
	 * @return
	 */
	public Connection getConexao() {
		return conexao;
	}
	/**
	 * Atribui a conexão com banco de dados. 
	 * @param conexao
	 */
	public void setConexao(Connection conexao) {
		this.conexao = conexao;
	}

	/**
	 * Define tipod e conexão para um DAO. 
	 * @param tipoDeConexao
	 */
	public void setTipoDeConexao(int tipoDeConexao) {
		this.tipoDeConexao = tipoDeConexao;
	}

	/**
	 * @return entidade
	 */
	public String getEntidade() {
		return entidade;
	}

	/**
	 * @param entidade 
	 */
	public void setEntidade(String entidade) {
		this.entidade = entidade;
	}

	/**
	 * @return the sgdb
	 */
	public String getSgdb() {
		return sgdb;
	}

	/**
	 * @param sgdb
	 */
	public void setSgdb(String sgdb) {
		this.sgdb = sgdb;
	}

	/**
	 * Arquivo de configuração do banco de dados. 
	 */
	public static final String ARQUIVO_CONFIGURACAO = "/dados/unicaffe/unicaffe_bd.ini";

	/**
	 * Tipo padrãod e banco de dados. 
	 */
	public static final int TIPO_DEFAULT = 0;
	/**
	 * Tipo de banco de dados de informações de usuários. 
	 */
	public static final int TIPO_USUARIOS = 1;
	/**
	 * Tipo de banco de dados para autenticação. 
	 */
	public static final int TIPO_AUTENTICACAO = 2;

	/**
	 * Drive jdbc para Sqlite. 
	 */
	public static final String DRIVER_SQLITE = "org.sqlite.JDBC";
	/**
	 * Banco de dados squlite
	 */
	
	public static final String JDBC_BANCO_SQLITE = "jdbc:sqlite:";

	/**
	 * JDBC para postgres. 
	 */
	public static final String JDBC_BANCO_POSTGRES = "jdbc:postgresql:";
	/**
	 * Driver JDBC postgres
	 */
	public static final String DRIVER_POSTGRES = "org.postgresql.Driver";
	/**
	 * JDBC Mysql
	 */
	public static final String JDBC_BANCO_MYSQL = "jdbc:mysql:";
	/**
	 * Driver JDBC Mysql
	 */
	public static final String DRIVER_MYSQL = "com.mysql.jdbc.Driver";

}
